gdpr/validate_block.go

```diff
@@ -0,0 +1,247 @@
+package gdpr
+
+import (
+	"errors"
+	"fmt"
+	"github.com/golang/protobuf/proto"
+
+	//"crypto"
+	"crypto/sha256"
+	"github.com/hyperledger/fabric-protos-go/common"
+	"github.com/hyperledger/fabric/core/ledger/kvledger/txmgmt/rwsetutil"
+	"github.com/hyperledger/fabric/protoutil"
+	//"hash"
+)
+
+var ErrVal = errors.New("gdpr: block validation failed")
+
+func getHash(preimage []byte) []byte{
+	h := sha256.New()
+	return h.Sum(preimage)
+
+}
+
+
+func validate (block *common.Block) (*common.Block , error){
+	preImages := block.GetData().GetPreimageSpace()
+
+	m := map[string]struct{}{}
+
+	for _, im := range preImages {
+		temp := getHash(im)
+		m[(string(temp))] = struct{}{}
+	}
+
+	for _ , envBytes := range block.Data.Data {
+		env, err := protoutil.GetEnvelopeFromBlock(envBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		payload, err := protoutil.UnmarshalPayload(env.Payload) //protoutil.GetBytesPayload()
+		if err != nil {
+			return nil, err
+		}
+
+		tx, err := protoutil.UnmarshalTransaction(payload.Data)
+		if err != nil {
+			return nil, err
+		}
+
+		_, respPayload, err := protoutil.GetPayloads(tx.Actions[0])
+		if err != nil {
+			return nil, err
+		}
+
+		txRWSet := &rwsetutil.TxRwSet{}
+		//txRWSet.FromProtoBytes(respPayload.Results)
+
+		if err = txRWSet.FromProtoBytes(respPayload.Results); err != nil {
+			return nil, err
+		}
+
+		for _, nsRWSet := range txRWSet.NsRwSets {
+			//nsRWSet.KvRwSet
+			for _, kvWrite := range nsRWSet.KvRwSet.Writes {
+				var a = kvWrite.GetValueHash()
+				if a != nil{
+					fmt.Printf("that's odd")
+				}
+				var b = getHash(kvWrite.GetValue())
+				//kvWrite.ValueHash = []byte("Gal")
+				if memberOf((string)(b),m) == false{
+					return nil,  ErrVal
+				}
+
+			}
+
+		}
+	}
+
+	return getVanillaBlock(block)
+
+	//return nil, nil
+}
+
+
+
+
+func memberOf(a string, m map[string]struct{}) bool {
+	_ , exists := m[a]
+	return exists
+
+}
+
+func getVanillaBlock(block *common.Block) (*common.Block , error) {
+	newBlock := proto.Clone(block).(*common.Block)
+	newBlock.Data.PreimageSpace = nil
+
+	clearKVWrites(newBlock) // TODO: remove when blocks are generated without KVWrite values.
```

Comment:
the `clearKVWrites` should be used in the test, not in the production code. It's fine to prepare the input in the test any way you like, however the production code should reflect the desired future construction